### PR TITLE
Feature: support symlinked /home

### LINKED ISSUE #14557: xz install fails on Fedora Silverblue
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
==> Fetching xz
Warning: Building xz from source as the bottle needs:
- HOMEBREW_CELLAR: /home/linuxbrew/.linuxbrew/Cellar (yours is /var/home/linuxbrew/.linuxbrew/Cellar)
- HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew (yours is /var/home/linuxbrew/.linuxbrew)
==> Downloading https://downloads.sourceforge.net/project/lzmautils/xz-5.4.1.tar.gz
```

(On Silverblue, /var/home is /home) :)

### Verification

- [x] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [x] I ran `brew update` and am still able to reproduce my issue.
- [x] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Install xz

### What happened (include all command output)?

See above

### What did you expect to happen?

For it to work.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
`brew install xz` on Fedora Silverblue
```

### PR DESCRIPTION
**Pull request by osalbahr**
Fixes #14557

This commit serves as a proof-of-concept that using Homebrew bottles is possible on Fedora Silverblue and other setups where `/home` is symlinked as`/home -> var/home`. The `install.sh` script does install in `/home/linuxbrew/.linuxbrew` it just so happens that its physical location is `/var/home/linuxbrew/.linuxbrew` due to the symlink. So, all programs are actually in the correct prefix. However due to the way `bin/brew` resolves symlinks, Homebrew commands think the prefix is custom. Me and @rohanssrao have tested that this fix also works in `fedora-toolbox` and `ubuntu` containers where we symlinked `/home` ->`var/home`. The script can be found at [`silverbrew-fedora-toolbox.sh`](https://github.com/osalbahr/brew/blob/silverbrew-develop/silverbrew-fedora-toolbox.sh) and you can directly run it in a new container after `podman run -it fedora-toolbox|ubuntu`. I am opening this PR to see if removing one `-P` causes unexpected regression. It seems to work fine on macOS too after this edit. To the best of my understanding, the goal is just to find the Homebrew directory from the `brew` symlink and it seems to achieve that goal without the removed `-P`. This change was originally proposed in Homebrew/discussions#1282 but was not tested to the best of my knowledge.

Before:
```console
[nebula@fedora Homebrew]$ brew install xz
==> Fetching xz
Warning: Building xz from source as the bottle needs:
- HOMEBREW_CELLAR: /home/linuxbrew/.linuxbrew/Cellar (yours is /var/home/linuxbrew/.linuxbrew/Cellar)
- HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew (yours is /var/home/linuxbrew/.linuxbrew)
==> Downloading https://downloads.sourceforge.net/project/lzmautils/xz-5.4.3.tar.gz
==> Downloading from https://netactuate.dl.sourceforge.net/project/lzmautils/xz-5.4.3.tar.gz
####################################################################################################################################################################################################################################### 100.0%
Error: The following formula cannot be installed from bottle and must be
built from source.
  xz
Install Clang or run `brew install gcc`.
```

After:
```console
[nebula@fedora Homebrew]$ git diff
diff --git a/bin/brew b/bin/brew
index afc02e9eb..373294806 100755
--- a/bin/brew
+++ b/bin/brew
@@ -40,7 +40,7 @@ do
 done
 unset cmd

-BREW_FILE_DIRECTORY="$(quiet_cd "${0%/*}/" && pwd -P)"
+BREW_FILE_DIRECTORY="$(quiet_cd "${0%/*}/" && pwd)"
 HOMEBREW_BREW_FILE="${BREW_FILE_DIRECTORY%/}/${0##*/}"
 HOMEBREW_PREFIX="${HOMEBREW_BREW_FILE%/*/*}"

[nebula@fedora Homebrew]$ brew install xz
==> Fetching xz
==> Downloading https://ghcr.io/v2/homebrew/core/xz/manifests/5.4.3
Already downloaded: /var/home/nebula/.cache/Homebrew/downloads/8d9f9cb1c34807246a161f74e3de29be172826599700e583c128fa2340fe177f--xz-5.4.3.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/xz/blobs/sha256:91c390f663972a9249fbbd98546bc2687afab99e5c740065d33cb474ed612bf5
Already downloaded: /var/home/nebula/.cache/Homebrew/downloads/06ce31cc5406b5ea435a2580360a65549816c1b5412f25fdb5dde5b136c90da9--xz--5.4.3.x86_64_linux.bottle.tar.gz
==> Pouring xz--5.4.3.x86_64_linux.bottle.tar.gz
🍺  /home/linuxbrew/.linuxbrew/Cellar/xz/5.4.3: 312 files, 3.7MB
==> Running `brew cleanup xz`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
```
Usage:
```console
[nebula@fedora test-xz]$ ls
[nebula@fedora test-xz]$ echo "Hello, Silverblue" > old.txt
[nebula@fedora test-xz]$ ls
old.txt
[nebula@fedora test-xz]$ xz old.txt
[nebula@fedora test-xz]$ ls
old.txt.xz
[nebula@fedora test-xz]$ file old.txt.xz
old.txt.xz: XZ compressed data, checksum CRC64
[nebula@fedora test-xz]$ xz -d old.txt.xz
[nebula@fedora test-xz]$ ls
old.txt
[nebula@fedora test-xz]$ cat old.txt
Hello, Silverblue
[nebula@fedora test-xz]$ xz --version
xz (XZ Utils) 5.4.3
liblzma 5.4.3
[nebula@fedora test-xz]$ which xz
/home/linuxbrew/.linuxbrew/bin/xz
```

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
Note: `brew style|typecheck|tests` fail with the following error message:
```console
An error occurred while installing json (2.6.3), and Bundler cannot
continue.

In Gemfile:
  rubocop-rspec was resolved to 2.20.0, which depends on
    rubocop-capybara was resolved to 2.18.0, which depends on
      rubocop was resolved to 1.50.2, which depends on
        json
Error: failed to run `/var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/bin/bundle install`!
[nebula@fedora ~]$ /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/bin/bundle install
/usr/bin/env: ‘ruby’: No such file or directory
[nebula@fedora ~]$ ruby
bash: ruby: command not found
```
But I think this might be unrelated to this commit. Fedora Silverblue 38 (latest) does not have `ruby` installed in the host.

-----
System info:
```console
[nebula@fedora Homebrew]$ brew config
HOMEBREW_VERSION: 4.0.28-17-gc4c79c2
ORIGIN: https://github.com/Homebrew/brew
HEAD: c4c79c25a3dafc9f22f8c05439d7b5ed4f286c6a
Last commit: 7 minutes ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b4f8dec4ac8cde0a20c6db1ca6c6d16df8e4c867
Core tap last commit: 7 hours ago
Core tap branch: master
Core tap JSON: 09 Jul 01:38 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_REPOSITORY: /var/home/linuxbrew/.linuxbrew/Homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: octa-core 64-bit skylake
Clang: N/A
Git: 2.40.0 => /bin/git
Curl: 7.87.0 => /bin/curl
Kernel: Linux 6.2.9-300.fc38.x86_64 x86_64 GNU/Linux
OS: Fedora release 38 (Thirty Eight)
Host glibc: 2.37
/usr/bin/gcc: N/A
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: 13.1.0
xorg: N/A
[nebula@fedora Homebrew]$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: No developer tools installed.
Install Clang or run `brew install gcc`.
```
----
Motivation/alternatives:
Since Fedora Silverblue follows the immutable paradigm (similarly to macOS), using Homebrew on Silverblue has the same added benefits as using Homebrew on macOS. The alternatives to this PR require using Homebrew in a container or a wrapper and I would like to reduce dependency on containers, or disabling selinux to create a permanent mount which is quite invasive.

### CODE REVIEW

**Comment by Bo98** [comment]
This makes sense and I agree that the Silverblue situation something that we should fix.

Although I agree that not resolving makes sense, I am however a little cautious about anyone relying on the reverse at this stage. Such a symlink setup would probably be quite crazy, but it has been over a decade of the original behaviour so I wouldn't put it past someone to have something like `ln -s /home/linuxbrew/.linuxbrew /usr/local` or similar. If the old behaviour was around for only a few months I would have been less cautious.

Given this is about using bottles, all that really matters is the default prefix. So we could have a middleground something like:

```bash
if [[ "${HOMEBREW_PREFIX}" != "${HOMEBREW_DEFAULT_PREFIX}" && "$(quiet_cd "${HOMEBREW_DEFAULT_PREFIX}" && pwd -P)" == "${HOMEBREW_PREFIX}" ]]
then
  HOMEBREW_PREFIX="${HOMEBREW_DEFAULT_PREFIX}"
fi
```

I would probably say though that any fix would ideally affect `HOMEBREW_REPOSITORY` too to be consistent.

**Comment by osalbahr** [comment]
@Bo98 thx! We actually thought about doing something along the lines of
```bash
if [[ $(realpath '/home/linuxbrew/.linuxbrew') == HOMEBREW_PREFIX ]]
then
   HOMEBREW_PREFIX='/home/linuxbrew/.linuxbrew'
fi
```
But your way looks more elegant/portable as it doesn't require `realpath`. I was initially hesitant to use `realpath` because I noticed that [the Homebrew codebase uses it as a custom function](https://github.com/Homebrew/install/blob/master/uninstall.sh#L80). I will amend the commit to use your fix instead.

**Comment by osalbahr** [comment]
It looks like `HOMEBREW_DEFAULT_PREFIX` isn't defined until [brew.sh#L39](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/brew.sh#L39). And at that point `HOMEBREW_REPOSITORY` has already been set in [brew#L53](https://github.com/Homebrew/brew/blob/master/bin/brew#L53). My knee jerk reaction is to also manually modify `HOMEBREW_REPOSITORY` in the same if statement in brew.sh, but I'd have to double check if `HOMEBREW_REPOSITORY` is the only variable that depends on `HOMEBREW_PREFIX`. Let me know if you intended for the if statement to be somewhere else.

**Comment by osalbahr** [comment]
I was thinking to add the if statement at [brew.sh#L659](https://github.com/Homebrew/brew/blob/4bbc0af6935816c75e0057364a84fd7fa8de071b/Library/Homebrew/brew.sh#L659). But at that point `HOMEBREW_CELLAR` has already been set.

This seems like it can be easily fixable by switching the order, i.e. set `HOMEBREW_DEFAULT_PREFIX`, correct for a symlinked `/home`, and then set `HOMEBREW_CELLAR` only afterwards. This only moves the first if statement in `brew.sh` further down a bit. I think this is safe since `HOMEBREW_CELLAR` doesn't get touched until [brew.sh#L67](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/brew.sh#L67) so might as well set the variable right before it's used. Thoughts?

**Comment by osalbahr** [comment]
2. Another fix, move the default prefix variables from `brew.sh` into `bin/brew` and add them to `FILTERED_ENV`. This also seems fine assuming `brew.sh` never gets called directly (which I don't think it gets called directly).

3. Instead of moving the variables and modifying `FILTERED_ENV`, we can simply copy the default prefix variables. I personally don't like duplicating code, as this requires changing stuff in two files, but I don't think changing the default prefix variables is something that happens often anyways. This one is the least intrusive but it's also the least elegant.

Solution (2) makes the most sense to me.

**Comment by osalbahr** [comment]
I finished editing and squashing commits. Both the CI and our local tests pass.

I kept the changes as as two commits for now. The relocation itself is harmless afaik. So keeping the changes as two commits might be useful for `git bisect`. Let me know if I should squash them together.

**Comment by rohanssrao** [comment]
To test our patch yourself, do the following to apply it on a machine that has `/home` symlinked to `var/home`:

```bash
cd $(brew --repository)
git checkout master
curl https://raw.githubusercontent.com/osalbahr/brew/silverbrew-develop/Library/Homebrew/brew.sh > /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.sh 
```

**Comment by Bo98** [comment]
Looks good from a quick glance. `brew config` returns as you would expect as well?

**Comment by osalbahr** [comment]
> Looks good from a quick glance. `brew config` returns as you would expect as well?

Yes, it did. I’ll paste the updated `brew config` later today.

**Comment by rohanssrao** [comment]
I believe so. `HOMEBREW_PREFIX` is set correctly.
```
HOMEBREW_VERSION: 4.0.28-20-g5b30de8-dirty
ORIGIN: https://github.com/Homebrew/brew
HEAD: 5b30de880d3c937b121dcc036582143dc4548b4f
Last commit: 24 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 7e1b8f2173dbc90b0b89d0777ca07159d732e95f
Core tap last commit: 22 hours ago
Core tap branch: master
Core tap JSON: 10 Jul 13:54 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: octa-core 64-bit zen
Clang: N/A
Git: 2.34.1 => /bin/git
Curl: 7.81.0 => /bin/curl
Kernel: Linux 6.3.8-200.fc38.x86_64 x86_64 GNU/Linux
OS: Unknown
Host glibc: 2.35
/usr/bin/gcc: N/A
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
```
This is on an Ubuntu 22.04 container with a symlinked home.

**Comment by osalbahr** [comment]
```console
$ brew config
HOMEBREW_VERSION: 4.0.28-29-g97ebbf3
ORIGIN: https://github.com/Homebrew/brew
HEAD: 97ebbf3f1f30dc1fcc7125edbafb9580e084db9d
Last commit: 4 minutes ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 42d44c493e037d543ef033e5c0d7444c2e5bf3a8
Core tap last commit: 35 minutes ago
Core tap branch: master
Core tap JSON: 10 Jul 19:17 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: octa-core 64-bit skylake
Clang: N/A
Git: 2.41.0 => /bin/git
Curl: 8.0.1 => /bin/curl
Kernel: Linux 6.3.11-200.fc38.x86_64 x86_64 GNU/Linux
OS: Fedora release 38 (Thirty Eight)
Host glibc: 2.37
/usr/bin/gcc: N/A
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: 13.1.0
xorg: N/A
```

- "Homebrew Ruby" starts with `/var`, but I don't think this is an issue.

```console
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: No developer tools installed.
Install Clang or run `brew install gcc`.
```

More OS info:
```
$ cat /etc/os-release
NAME="Fedora Linux"
VERSION="38.20230709.0 (Silverblue)"
ID=fedora
VERSION_ID=38
VERSION_CODENAME=""
PLATFORM_ID="platform:f38"
PRETTY_NAME="Fedora Linux 38.20230709.0 (Silverblue)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:38"
DEFAULT_HOSTNAME="fedora"
HOME_URL="https://silverblue.fedoraproject.org"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora-silverblue/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://github.com/fedora-silverblue/issue-tracker/issues"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=38
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=38
SUPPORT_END=2024-05-14
VARIANT="Silverblue"
VARIANT_ID=silverblue
OSTREE_VERSION='38.20230709.0'
```

**Comment by Bo98** [comment]
> "Homebrew Ruby" starts with /var, but I don't think this is an issue.

Yeah, that's fine.

**Review by Bo98 — APPROVED** [review]

**Review by iMichka — APPROVED** [review]

**Comment by iMichka** [comment]
Thanks for working on a proper fix! Let's see how this goes :)

**Comment by osalbahr** [comment]
Thx! I just tested on a Fedora Silverblue VirtualBox VM and it seems to work fine.
```console
[silverblue-vbox@fedora ~]$ brew list
xz
[silverblue-vbox@fedora ~]$ brew developer
Developer mode is enabled.
[silverblue-vbox@fedora ~]$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: No developer tools installed.
Install Clang or run `brew install gcc`.
[silverblue-vbox@fedora ~]$ brew config
HOMEBREW_VERSION: 4.0.28-70-g518fb43
ORIGIN: https://github.com/Homebrew/brew
HEAD: 518fb43112a7853c3476b0789ce3da23fb752cb6
Last commit: 11 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: a3267a34c4be8286044ab9aad1d98e685bbd63ab
Core tap last commit: 2 hours ago
Core tap branch: master
Core tap JSON: 15 Jul 22:16 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_MAKE_JOBS: 1
Homebrew Ruby: 2.6.10 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: single-core 64-bit skylake
Clang: N/A
Git: 2.40.0 => /bin/git
Curl: 7.87.0 => /bin/curl
Kernel: Linux 6.2.9-300.fc38.x86_64 x86_64 GNU/Linux
OS: Fedora release 38 (Thirty Eight)
Host glibc: 2.37
/usr/bin/gcc: N/A
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
```

File: Library/Homebrew/brew.sh [hunk 1]
```diff
@@ -55,6 +44,37 @@ else
   HOMEBREW_DEFAULT_TEMP="/tmp"
 fi
 
+realpath() {
+  (cd "$1" &>/dev/null && pwd -P)
+}
+
+# Support systems where HOMEBREW_PREFIX is the default,
+# but a parent directory is a symlink.
+# Example: Fedora Silverblue symlinks /home -> var/home
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:53]: @osalbahr @iMichka should this say `/var/home`?

Comment (iMichka) [reply Library/Homebrew/brew.sh:53]: I think yes, that's a typo in the comment. @osalbahr up for a follow up PR?

Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:53]: Great, thanks!

**Pull request merged**